packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +424,411 @@ contract AmmPool is IBlockReceiver {
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].add(amount);
+                    address token = ctx.tokens[i].addr;
+                    balance[token][join.owner] = balance[token][join.owner].sub(amount);
                 }
+                ctx.ammBalances[i] = ctx.ammBalances[i].add(amount);
             }
-            // Clean up the queue item to get the gas refund
-            delete queue[n];
+
+            // Mint liquidity tokens
+            mint(join.owner, join.poolAmountOut);
         }
-        queuePos += auxData.numItems;
 
-        // Deposit/Withdraw to/from the AMM account when necessary
-        for (uint i = 0; i < tokens.length; i++) {
-            if (ammBalances[i] > ammBalancesInAccount[i]) {
-                uint amount = ammBalances[i] - ammBalancesInAccount[i];
-                // Check that the deposit in the block matches the expected deposit
-                DepositTransaction.Deposit memory _deposit = _block.readDeposit(txIdx++);
-                require(_deposit.owner == address(this), "INVALID_TX_DATA");
-                require(_deposit.accountID == accountID, "INVALID_TX_DATA");
-                require(_deposit.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
-                require(_deposit.amount == amount, "INVALID_TX_DATA");
-                // Now do this deposit
-                uint ethValue = 0;
-                if (tokens[i].addr == address(0)) {
-                    ethValue = _deposit.amount;
+        if (!join.fromLayer2) {
+            for (uint i = 0; i < ctx.tokens.length; i++) {
+                address token = ctx.tokens[i].addr;
+                uint amount = join.maxAmountsIn[i];
+                // Unlock the amount locked for this join
+                locked[token][join.owner] = locked[token][join.owner].sub(amount);
+            }
+        }
+    }
+
+    function processExit(
+        Context  memory ctx,
+        PoolExit memory exit,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolExit(ctx.DOMAIN_SEPARATOR, exit);
+        if (signature.length == 0) {
+            require(queue[queuePos].txHash == poolTxHash, "NOT_APPROVED");
+            delete queue[queuePos];
+            queuePos++;
+        } else {
+            require(poolTxHash.verifySignature(exit.owner, signature), "INVALID_SIGNATURE");
+        }
+
+        uint poolTotal = totalSupply();
+        uint ratio = (exit.poolAmountIn * BASE) / poolTotal;
+
+        // Check if the requirements are fulfilled
+        bool valid = availableBalance(address(this), exit.owner) >= exit.poolAmountIn;
+        uint[] memory amounts = new uint[](ctx.tokens.length);
+        for (uint i = 0; i < ctx.tokens.length; i++) {
+            amounts[i] = ctx.ammBalances[i] * ratio / BASE;
+            if(amounts[i] < exit.minAmountsOut[i]) {
+                valid = false;
+            }
+        }
```

Comment:
I suggest we extra a `validateExit` method and returns a boolean for 'valid'. Inside `validateExit`, we should also check there are enough liquidity tokens to burn then burn it there.